File: internal/cli/shared.go
package cli

import (
	"errors"
	"fmt"
	"path/filepath"
	"strings"

	"github.com/Slothtron/mwt/internal/config"
	"github.com/Slothtron/mwt/internal/pathresolve"
)

// selectRepos returns the configured repos to operate on.
// When requested is empty, all cfg.Repos are used.
// Unknown names are an error.
func selectRepos(cfg *config.Config, requested []string) ([]string, error) {
	if cfg == nil {
		return nil, fmt.Errorf("config is nil")
	}
	if len(requested) == 0 {
		return append([]string(nil), cfg.Repos...), nil
	}

	known := make(map[string]struct{}, len(cfg.Repos))
	for _, r := range cfg.Repos {
		known[r] = struct{}{}
	}

	seen := make(map[string]struct{}, len(requested))
	out := make([]string, 0, len(requested))
	for _, r := range requested {
		r = strings.TrimSpace(r)
		if r == "" {
			return nil, fmt.Errorf("empty repo name in --repos")
		}
		if _, ok := known[r]; !ok {
			return nil, fmt.Errorf("repo %q is not in config repos", r)
		}
		if _, dup := seen[r]; dup {
			continue
		}
		seen[r] = struct{}{}
		out = append(out, r)
	}
	return out, nil
}

// runSerial executes fn for each repo in order.
// Without continueOnErr, the first error stops the loop.
// With continueOnErr, remaining repos still run; any failure yields a joined
// non-nil error at the end (exit non-zero).
func runSerial(repos []string, continueOnErr bool, fn func(repo string) error) error {
	var errs []error
	for _, repo := range repos {
		if err := fn(repo); err != nil {
			wrapped := fmt.Errorf("%s: %w", repo, err)
			if !continueOnErr {
				return wrapped
			}
			errs = append(errs, wrapped)
		}
	}
	return errors.Join(errs...)
}

func resolveRepo(cfg *config.Config, repo, branch string) (pathresolve.Context, error) {
	return pathresolve.ResolveFromConfig(cfg, repo, branch)
}

func mainPath(cfg *config.Config, repo string) string {
	return filepath.Join(cfg.MetaRoot, repo)
}

File: src/cli/shared.rs
//! Cross-command helpers shared by the CLI subcommand implementations.

use std::path::{Path, PathBuf};

use crate::config::Config;
use crate::pathresolve;

/// `selectRepos`: pick the subset of `cfg.repos` named in `requested`.
/// Empty `requested` → all repos. Unknown names → error.
pub fn select_repos(cfg: &Config, requested: &[String]) -> Result<Vec<String>, String> {
    let known: std::collections::HashSet<&str> = cfg.repos.iter().map(String::as_str).collect();
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut out: Vec<String> = Vec::new();
    for r in requested {
        let r = r.trim();
        if r.is_empty() {
            return Err("empty repo name in --repos".to_string());
        }
        if !known.contains(r) {
            return Err(format!("repo {r:?} is not in config repos"));
        }
        if seen.insert(r.to_string()) {
            out.push(r.to_string());
        }
    }
    if out.is_empty() {
        Ok(cfg.repos.clone())
    } else {
        Ok(out)
    }
}

/// `runSerial`: invoke `fn` for each repo in order.
/// `continue_on_err=true` keeps going and joins all errors at the end.
pub fn run_serial<F>(repos: &[String], continue_on_err: bool, mut fn_: F) -> Result<(), String>
where
    F: FnMut(&str) -> Result<(), String>,
{
    let mut errs: Vec<String> = Vec::new();
    for repo in repos {
        if let Err(e) = fn_(repo) {
            let wrapped = format!("{repo}: {e}");
            if !continue_on_err {
                return Err(wrapped);
            }
            errs.push(wrapped);
        }
    }
    if errs.is_empty() {
        Ok(())
    } else {
        Err(errs.join("; "))
    }
}

pub fn resolve_repo(
    cfg: &Config,
    repo: &str,
    branch: &str,
) -> Result<pathresolve::Context, String> {
    pathresolve::Context::resolve_from_config(cfg, repo, branch).map_err(|e| e.to_string())
}

pub fn main_path(cfg: &Config, repo: &str) -> PathBuf {
    cfg.meta_root.join(repo)
}

/// `displayPath`: prefer path relative to `meta_root`; fall back to absolute
/// when the path escapes `meta_root` or the relative computation fails.
pub fn display_path(meta_root: &Path, abs: &str) -> String {
    if meta_root.as_os_str().is_empty() || abs.is_empty() {
        return abs.to_string();
    }
    let abs_p = Path::new(abs);
    match abs_p.strip_prefix(meta_root) {
        Ok(rel) => {
            let s = rel.to_string_lossy();
            if s.is_empty() {
                return ".".to_string();
            }
            s.replace(std::path::MAIN_SEPARATOR, "/")
        }
        Err(_) => abs.to_string(),
    }
}

pub fn max_repo_width(repos: &[String]) -> usize {
    repos.iter().map(String::len).max().unwrap_or(0)
}
